### ISSUE TITLE
Usage of sigsetjmp/siglongjmp leads to undefined symbol references

### ISSUE THREAD

**Issue by atrosinenko** [kripken/emscripten#5204]
Currently, one may successfully compile a C source that references `sigsetjmp`/`siglongjmp` since they are defined in `include/libc/setjmp.h`:

```c
#if defined(_POSIX_SOURCE) || defined(_POSIX_C_SOURCE) \
 || defined(_XOPEN_SOURCE) || defined(_GNU_SOURCE) \
 || defined(_BSD_SOURCE)
typedef jmp_buf sigjmp_buf;
int sigsetjmp (sigjmp_buf, int);
_Noreturn void siglongjmp (sigjmp_buf, int);
#endif
```

But it leads to references to undefined symbols that crash program at run time:
```c
#include <setjmp.h>
#include <stdio.h>

int main() {
    sigjmp_buf env;
    if(sigsetjmp(env, 1) == 0) {
        fprintf(stderr, "Jumping...\n");
        siglongjmp(env, 1);
    } else {
        fprintf(stderr, "Jumped.\n");
    }
    return 0;
}
```
```
$ emcc test.c -o test.js
warning: unresolved symbol: siglongjmp
warning: unresolved symbol: sigsetjmp
$ node test.js
missing function: sigsetjmp
-1
-1
exception thrown: abort(-1) at Error
    at jsStackTrace (/tmp/bug/test.js:1125:13)
    at stackTrace (/tmp/bug/test.js:1142:12)
    at abort (/tmp/bug/test.js:9410:44)
    at _sigsetjmp (/tmp/bug/test.js:1775:54)
    at _main (/tmp/bug/test.js:1983:11)
    at Object.asm._main (/tmp/bug/test.js:9062:19)
    at Object.callMain (/tmp/bug/test.js:9284:30)
    at doRun (/tmp/bug/test.js:9348:60)
    at run (/tmp/bug/test.js:9362:5)
    at Object.<anonymous> (/tmp/bug/test.js:9436:1)

/tmp/bug/test.js:110
      throw ex;
      ^
abort(-1) at Error
    at jsStackTrace (/tmp/bug/test.js:1125:13)
    at stackTrace (/tmp/bug/test.js:1142:12)
    at abort (/tmp/bug/test.js:9410:44)
    at _sigsetjmp (/tmp/bug/test.js:1775:54)
    at _main (/tmp/bug/test.js:1983:11)
    at Object.asm._main (/tmp/bug/test.js:9062:19)
    at Object.callMain (/tmp/bug/test.js:9284:30)
    at doRun (/tmp/bug/test.js:9348:60)
    at run (/tmp/bug/test.js:9362:5)
    at Object.<anonymous> (/tmp/bug/test.js:9436:1)
```

According to `sigsetjmp`/`siglongjmp` man pages, they only differ from non-`sig` versions in predictable behavior considering signal masks. It looks like Emscripten currently does not support signals anyway, so they may be made just aliases to the non-`sig` versions.

I want to submit a PR and I think that since `setjmp` is some very low-level function that may be known to compiler, it may be worth just `#define`ing `sigsetjmp` to it somewhere in the headers (and adding test that `sizeof jmp_buf == sizeof sigjmp_buf`), but where is the right place: directly in the musl libc source code or somewhere in Emscripten-specific patches?

**Comment by kripken** [comment]
I worry a little that defining these could be confusing because we don't support signals. But I do see the point that it makes sense to implement them too. So probably fine to do this if it's useful.

If there is a good way to do it in the musl headers, that might be best. The optimal thing is if it could be done in our extra files for emscripten inside the musl headers. If it requires changes to core musl headers, if it's small enough maybe that's ok. Otherwise, another option is in fastcomp's setjmp handling pass to also detect sigsetjmp there, should be just a few lines.

**Comment by juj** [comment]
We do already implement deferred pthread cancellation. The same architecture could be used to implement support for deferred signal handling, i.e. cooperatively dispatch signals on the already defined thread cancellation points. I think this would be quite good enough for porting most signal using applications. This is not currently on the roadmap, but if you do want to submit PRs with implementation and tests towards this, those would be most welcome.

**Comment by atrosinenko** [comment]
Oh, I have probably incorrectly explained. Implementing signals was not on my roadmap too :) I just talked about implementing kind of no-op (with respect to signals) stub for `sigsetjmp` like stubs we currently have for `kill`, etc. Another option will probably be to **completely remove** definitions of unimplemented `sigsetjmp`/`siglongjmp` so configure script can detect their absence with more probability. And this should be much less confusing considering signal support.

**Comment by atrosinenko** [comment]
Implemented trivial stubs in #5295. I hope they will not confuse anyone since they should be explicitly enabled via `-D_EMSCRIPTEN_SIGLONGJMP_STUB`.

**Comment by atrosinenko** [comment]
I close this issue because although signals are still not implemented, the original issue was fixed in #5295 . The `multithreading` label can probably be dropped.

**Comment by juj** [comment]
Makes sense, thanks for the PR! I'll keep the multithreading label since they are often useful when I search for discussions that have related to a particular feature, so good to keep the label around.
